src/mbgl/style/expression/number_format.cpp

```diff
@@ -0,0 +1,276 @@
+                           optional<std::unique_ptr<Expression>> maxFractionDigits_)
+    : Expression(Kind::NumberFormat, type::String),
+      number(std::move(number_))
+{
+      if (locale_)
+            locale = std::unique_ptr<Expression>(std::move(*locale_));
+
+      if (currency_)
+            currency = std::unique_ptr<Expression>(std::move(*currency_));
+
+      if (minFractionDigits_)
+            minFractionDigits = std::unique_ptr<Expression>(std::move(*minFractionDigits_));
+
+      if (maxFractionDigits_)
+            maxFractionDigits = std::unique_ptr<Expression>(std::move(*maxFractionDigits_));
+}
+
+EvaluationResult NumberFormat::evaluate(const EvaluationContext& params) const {
+    auto numberResult = number->evaluate(params);
+    if (!numberResult) {
+        return numberResult.error();
+    }
+    double evaluatedNumber = numberResult->get<double>();
+    
+    optional<std::string> evaluatedLocale;
+    if (locale) {
+        auto localeResult = (*locale)->evaluate(params);
+        if (!localeResult) {
+            return localeResult.error();
+        }
+        evaluatedLocale = toString(*localeResult);
+        // Replace '-' with '_' in the locale string and append '.utf8' to it
+        boost::replace_first(*evaluatedLocale, "-", "_");
+        (*evaluatedLocale).append(".utf8");
+    }
+
+    optional<std::string> evaluatedCurrency;
+    if (currency) {
+        auto currencyResult = (*currency)->evaluate(params);
+        if (!currencyResult) {
+            return currencyResult.error();
+        }
+        evaluatedCurrency = toString(*currencyResult);
+    }
+
+    optional<double> evaluatedMinFractionDigits;
+    if (minFractionDigits) {
+        auto minDigitsResult = (*minFractionDigits)->evaluate(params);
+        if (!minDigitsResult) {
+            return minDigitsResult.error();
+        }
+        evaluatedMinFractionDigits = minDigitsResult->get<double>();
+    }
+
+    optional<double> evaluatedMaxFractionDigits;
+    if (maxFractionDigits) {
+        auto maxDigitsResult = (*maxFractionDigits)->evaluate(params);
+        if (!maxDigitsResult) {
+            return maxDigitsResult.error();
+        }
+        evaluatedMaxFractionDigits = maxDigitsResult->get<double>();
+    }
+
+    std::stringstream ss;
+    std::locale loc = std::locale(evaluatedLocale ? *evaluatedLocale : "");
+    ss.imbue(loc);
+
+    // Print the value as currency
+    if (evaluatedCurrency) {
+        const auto& mp = std::use_facet<std::moneypunct<char> >(loc);
+        int fracDigits = mp.frac_digits();
+        // Multiply with 10^fracDigits to print the value correctly
+        if (fracDigits) {
+            evaluatedNumber = evaluatedNumber * pow(10, fracDigits);
+        }
+        ss << std::showbase << std::put_money(evaluatedNumber);
+    } else {
+        int minPrecision = 0, maxPrecision = 3; //defaults
+        int integerPart = (int)evaluatedNumber;
```

Comment:
`long`? otherwise a insanely large double will overflow this

Reply:
btw, we should add a unit test for negative numbers

Reply:
Changed to ```long```